runners/core-java/src/main/java/org/apache/beam/runners/core/StatefulDoFnRunner.java

```diff
@@ -105,7 +105,7 @@ public void processElement(WindowedValue<InputT> input) {
 
   private boolean isLate(BoundedWindow window) {
     Instant gcTime = LateDataUtils.garbageCollectionTime(window, windowingStrategy);
-    Instant inputWM = cleanupTimer.currentInputWatermarkTime();
+    Instant inputWM = cleanupTimer.currentOutputWatermarkTime();
```

Comment:
Hmm, this seems like a very serious and potentially incompatible change; it needs to be validated at least by ValidatesRunner tests for other runners. I don't know enough about this code to tell for sure whether it is correct - hopefully @kennknowles can comment.

Generally, I doubt that this is correct. Output watermark can be quite decoupled from input watermark; and in theory there should be different watermarks for different output tags, i.e. there should be no single "output watermark". It feels very icky to use it for determining whether the input is late.

Reply:
I had an alternative idea which is setting an event-time timestamp of the fired processing-time timer that is not filtered out as late. It seems there are some complications with this as well but I think I'll figure it out. I can't work on it this week and next because I'm traveling, though.